// emsfdc/test/testproj/test/imageUpload/index.js
var util = require('util');
var fs = require('fs');
var async = require('async');

var session;

var uniqueName = 'Mark' + new Date().toString();

module.exports = function(sessionParam,cb) {
    session = sessionParam;
    selectAll();
}

function selectAll(err,res) {
    session.directive(
        {op: "SELECT", targetType:"Fairyland__c", properties:["id","number_of_guests__c","name_of_visitor__c","time_of_visit__c"]}
    ,deleteResultList);
}

function deleteResultList(err,result) {
    console.dir(result);
    if(result.results.length > 0) {
        console.log('Found ' + result.results.length + ' rows in Fairyland__c');
        async.forEachSeries(result.results,
            function(row,cb) {
                var request = {op: "DELETE",targetType: "Fairyland__c",where: {externalId: row.id}};
                console.dir(request);
                session.directive(request,cb);
            },
        insert1);
    } else {
        insert1();
    }
}

function insert1() {
    console.log("Inserting, name is " + uniqueName);
    session.directive({
        op:'INSERT',
        targetType: 'Fairyland__c', 
        values: {
            time_of_visit__c: '2011-10-31',
            name_of_visitor__c: uniqueName,
            number_of_guests__c: '2',
        }
    },wait1);
}

function wait1(err,res) {
    console.log('waiting...');
    setTimeout(select1,2000);
}

function select1(res) {
    session.directive({
        op: 'SELECT',
        targetType: 'Fairyland__c',
        properties:["id","number_of_guests__c","name_of_visitor__c","time_of_visit__c"], 
        where: {name_of_visitor__c:uniqueName}
    },uploadImage);
}

var resourceName;
var mongoId;

function uploadImage(err,res) {
    console.dir(res);

    mongoId = res.results[0].externalId;
    if(!mongoId) {
        console.log('FAILURE: Object was not created in Salesforce!');
        return;
    }

    var data = fs.readFileSync('icon.gif');
    var folderId = 'photos';
    var documentName = 'pic' + new Date().getTime();
    
    resourceName = '/' + folderId + '/' + documentName;

    session.directive({
        op: 'INVOKE',
        targetType: 'Document',
        name: 'refresh',
        params: {
            name: resourceName,
            folderId: 'photos',
            contentType: 'image/gif',
            content: data.toString('base64')
        }         
    },update1);

    console.log("Uploading image: " + resourceName);
}

function update1(err,res) {
    console.log("Upload complete");
     console.log("Updating...");

    session.directive({
        op:'UPDATE',
        targetType: 'Fairyland__c', 
        values: {
            photo_name__c: resourceName,
        },
        where : {externalId: mongoId}
    },download);
 }

function download(err,res) {
    console.log('downloading ' + uniqueName);
    session.directive({
        op: 'SELECT',
        targetType: 'Fairyland__c',
        properties:["id","number_of_guests__c","name_of_visitor__c","time_of_visit__c","photo_name__c"],
        where: {name_of_visitor__c:uniqueName}
    },getImage);
}

function getImage(err,res) {
    console.dir(res)

    var resourcePath = res.results[0].photo_name__c;

    console.log('Resource to download = ' + resourcePath);

    var part = resourcePath.split('/');

    session.directive({
        op: 'INVOKE',
        targetType: 'Document',
        name: 'fetch',
        params: {
            name: resourcePath
        }         
    },done);

}

function done(err,res) {
    console.dir(res);
    console.log('Done!');
    console.log('unique name = ' + uniqueName);
}
